### PR TITLE
Add `ReplayEventStream` for replaying events while using constant memory

### PR DESCRIPTION
**Pull request by kmadsen**
### Description

There has been a known issue with replaying history and routes, which is that it loads all the events into memory before starting the simulation and that will cause memory issues when there is a lot of data. Years ago I started solving this https://github.com/mapbox/mapbox-navigation-android/pull/3806. The reason I did not merge that, was because nav-native was adding protobuf history files and that was changing the details. I didn't pick it up again because we are not replaying sensor data. High frequency sensor data was the reason for large history files.

Now with the introduction of simulating routes with high frequency location updates https://github.com/mapbox/mapbox-navigation-android/pull/5748. There is a new need to handle large data for the simulations. Simulating a large route at high frequency requires a lot of memory and can cause OutOfMemory exceptions.

So this effort will be a 2 part series. 1 optimize history replay. 2 optimize route replay. This pull request is part 1 because it is simpler than optimizing the simulation algorithm.

Opening as draft because the `ReplayHistoryActivity` is all working as expected. But I'm still writing tests.

### Screenshots or Gifs

Introducing the new `ReplayEventStream`. This allows you to keep a low memory footprint while simulating events.

### CODE REVIEW

File: libnavigation-core/api/current.txt [hunk 1]
```diff
@@ -251,10 +251,8 @@ package com.mapbox.navigation.core.history {
 
   public final class MapboxHistoryReader implements java.util.Iterator<com.mapbox.navigation.core.history.model.HistoryEvent> kotlin.jvm.internal.markers.KMappedMarker {
     ctor public MapboxHistoryReader(String filePath);
-    method public String getFilePath();
```
Comment (Guardiola31337) [inline libnavigation-core/api/current.txt:254]: This seems like a breaking change. Should we expose a getter for `filePath` to not break SemVer?

**Review by Guardiola31337 — COMMENTED** [review]
Took a pass to the code and looks good overall (left a comment about SemVer in `MapboxHistoryReader`). Will test locally later.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/history/MapboxHistoryReader.kt [hunk 2]
```diff
@@ -1,50 +1,55 @@
...
+import com.mapbox.navigator.HistoryReaderInterface
 
 /**
  * Allows you to read history files previously saved by [MapboxHistoryRecorder].
  * All files in the [MapboxHistoryRecorder.fileDirectory] can be read with this reader.
- *
- * @param filePath absolute path to a file containing the native history file.
  */
-class MapboxHistoryReader(
-    val filePath: String
```
Comment (kmadsen) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/history/MapboxHistoryReader.kt:14]: ah yes here needs to be preserved

File: libnavigation-core/api/current.txt (hunk 1 above)
Comment (kmadsen) [reply libnavigation-core/api/current.txt:254]: yep nice catch! thank you

Comment (kmadsen) [reply libnavigation-core/api/current.txt:254]: https://github.com/mapbox/mapbox-navigation-android/pull/6582/files#r1023383239

**Comment by kmadsen** [comment]
Closing this one because the approach was not reliable. Next iteration I think we should do something similar to `ReplayRouteSession`  https://github.com/mapbox/mapbox-navigation-android/pull/6636

So it would be called `ReplayHistorySession` cc: @VysotskiVadim

**Pull request closed without merging**